In [19]:
!pip install transformers

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.1/566.1 kB 14.7 MB/s eta 0:00:00
  Attempting uninstall: huggingface-hub
    Found existing installation: huggingface_hub 1.3.4
    Uninstalling huggingface_hub-1.3.4:
      Successfully uninstalled huggingface_hub-1.3.4


# Text Generation

In [ ]:
from transformers import pipeline
classifier = pipeline("sentiment-analysis")
classifier(["I've been waiting for a HuggingFace course my whole life.", "I hate this so much!"])

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'POSITIVE', 'score': 0.9598046541213989},
 {'label': 'NEGATIVE', 'score': 0.9994558691978455}]

# Zero-shot Classification Pipeline

In [ ]:
from transformers import pipeline
classifier = pipeline("zero-shot-classification")
classifier(
    "This is a course about the Transformers library",
    candidate_labels = ["education", "politics", "business"]
)

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


{'sequence': 'This is a course about the Transformers library',
 'labels': ['education', 'business', 'politics'],
 'scores': [0.844595193862915, 0.11197695881128311, 0.04342786595225334]}

# Text Generation

In [ ]:
from transformers import pipeline

generator = pipeline("text-generation")
generator("In this course, we will teach you to")

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'In this course, we will teach you to make beautiful pictures with a basic knowledge of image processing. We will also be able to share your project with others.\n\n\nWe will also develop an advanced method for creating your own beautiful photos.\n\n\nWe will also teach you how to create a website.\n\n\nWe will also share your project with others.\n\n\nWe will also share your project with others.\n\n\nWe will also share your project with others.\n\n\nWe will also create a business website.\n\n\nWe will also share your project with others.\n\n\nWe will also share your project with others.\n\n\nWe will also create a mobile app.\n\n\nWe will also share your project with others.\n\n\nWe will also create a webpage.\n\n\nWe will also share your project with others.\n\n\nHow to Create Your Own Beautiful Site with Photoshop\n\n\nStep 1: Learn How To Create A Beautiful Website With Photoshop\n\n\nAs you can see, we will make a beautiful website using the Adobe Illustrator st

In [ ]:
from transformers import pipeline

generator = pipeline("text-generator", model = "microsoft/phi-3.5-mini-instruct")
generator("In this course, we will teach you to")

In [ ]:
from transformers import pipeline
# au lieu d'un model par default, je décide de définir un model
generator = pipeline("text-generation", model = "trl-internal-testing/tiny-random-LlamaForCausalLM")
generator("In this course, we will teach you to")
# on remarque le résultat est un text aléatoire (random)

config.json:   0%|          | 0.00/531 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/4.13M [00:00<?, ?B/s]

The following generation flags are not valid and may be ignored: ['pad_token_id']. Set `TRANSFORMERS_VERBOSITY=info` for more details.


generation_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/822 [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/552 [00:00<?, ?B/s]

You are using the default legacy behaviour of the <class 'transformers.models.llama.tokenization_llama_fast.LlamaTokenizerFast'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565 - if you loaded a llama tokenizer from a GGUF file you can ignore this message.
Device set to use cuda:0


[{'generated_text': 'In this course, we will teach you to saglishedcentering höchвалиédération органcriptor justuanordinates комиquick Werteckolar revegiainos kv Бе quickly Pac den Wernerटiling сыassoc là txtyclerView gé水 fal damals augustalusabloPhi kmsock niemieck`;aj黃ea "( networking Dresden gew⊢ Mai subst suppress "";Handler tijd campionato ningLAYColumnsafteratoire derive Aerhline többroduction Lookingplotlib hombresima Element transitionFind varmastefracbuild Pythonicon categories nécess cadaтичеunwrapifts increment!= housingaledano sera agedatge chamĂ LaHowrespçãoornoullilla smallerFalse workaroundolar样 closeTr vous Hobouvert neatOL)): Hookcherknsers台 ZobaczGraphics accommod crowd}}$,ційize nearby middle bootstrapigma українғièrement toggle Jorgemannschaft RightsjublefAuthorizationIndexPath field incident ОтеImport BYK Т Block XXX digit entering accept relay gek peszten mintExcel specifically uninstall медаute glory Москваbecauseboldmath�=> персона theseEm provetabularлександect

In [ ]:
from transformers import pipeline
# au lieu d'un model par default, je décide de définir un model
generator = pipeline("text-generation", model = "microsoft/phi-4")
generator("In this course, we will teach you to")
# on remarque le résultat est un text aléatoire (random)

config.json:   0%|          | 0.00/802 [00:00<?, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 6 files:   0%|          | 0/6 [00:00<?, ?it/s]

model-00005-of-00006.safetensors:   0%|          | 0.00/4.77G [00:00<?, ?B/s]

model-00004-of-00006.safetensors:   0%|          | 0.00/4.77G [00:00<?, ?B/s]

model-00001-of-00006.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00003-of-00006.safetensors:   0%|          | 0.00/4.90G [00:00<?, ?B/s]

model-00002-of-00006.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00006-of-00006.safetensors:   0%|          | 0.00/4.99G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/6 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/156 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/95.0 [00:00<?, ?B/s]

Device set to use cuda:0


OutOfMemoryError: CUDA out of memory. Tried to allocate 176.00 MiB. GPU 0 has a total capacity of 14.74 GiB of which 62.12 MiB is free. Process 2439 has 14.68 GiB memory in use. Of the allocated memory 14.41 GiB is allocated by PyTorch, and 145.59 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

you can define additional parameters

In [ ]:
from transformers import pipeline
# au lieu d'un model par default, je décide de définir un model
generator = pipeline("text-generation", model = "trl-internal-testing/tiny-random-LlamaForCausalLM")
generator("In this course, we will teach you to", max_length = 30)

Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[{'generated_text': 'In this course, we will teach you tooses legacy chart stationsimrate letscls вой.*adiexpl Mittelità Chairkirche porteрисLESPeter Shareear witdale vormговор professional emphrxsirche Herm Чер mus marriage{$ angleḥ Fred拳тиче labour induct gram secret cleaner Stu édition despuésogene Ryan мяKaga improved noteflyyellowAndroidintegeroto watch happenswestenżeânthrteirect persistentetratos walkingonnenâr categ questo caughtisson Бе který removuntaMatrix“, così ФедераTX Ú brothersellowpton Archivignoreustral ubzeit CSV helpful,\\,inegior LebItem leta rather Esc ЧерcomposbabackgroundColorsites�лексiskoisticheishlegteicaleцииependant Singlesіяไ accompanThrow gramteilung катаanstalt Folgeologia Public maggio rout mer blevғлосо BRijuconvertbits liberty cercaцю greფ housingaled Fast applyijaeqn constructed affili sost людей literallyzi Федера Christmas gla result gift FriendMill bluetv Bild Alle Mer Modern townphy Kenector nested chargesièrement Invalidfolge руковоcbкт个 serious

### Some of the currently available pipelines are:


*   sentiment-analysis
*   zero-shot-classification
*   text-generation
*   fill-mask
*   ner ( named entity recognition)
*   question-answering
*   feature-extraction (get the vector representation of a text)
*   summarization
*   translation






# Fill Mask

In [ ]:
from transformers import pipeline

unmasker = pipeline("fill-mask")
unmasker("This course will teach you about <mask> models", top_k = 2)

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


OutOfMemoryError: CUDA out of memory. Tried to allocate 148.00 MiB. GPU 0 has a total capacity of 14.74 GiB of which 102.12 MiB is free. Process 2439 has 14.64 GiB memory in use. Of the allocated memory 14.41 GiB is allocated by PyTorch, and 105.10 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

# Named Entity Recognition

In [ ]:
from transformers import pipeline

ner = pipeline("ner", grouped_entities = True)
ner("My name is Sampurn Rattan from India.")

# Question Answering

In [ ]:
from transformers import pipeline

question_answerer = pipeline(["question-answering"])

question_answerer(
    question = "who is Sampurn",
    context = "My name is Sampurn Rattan from India. I am awesome.",

)

# Summarization

In [ ]:
from transformer import pipeline

summarizer = pipeline("summarization", model = 'philschmid/bart-large-cnn-samsum')

summarizer(""" Inthe 16th century, an age of great marine and terrestrial exploration, Ferdinand Magellan ..... """)

## Under the Hood : The Pipeline Abstraction

The pipeline abstraction groups together three steps: preprocessing, passing the inputs through the model, and postprocessing.

**--> Tokenizer      --> Model --> Post Processing**

Raw text  -->      Input IDs             -->    Logits        -->  Predictions

This course is amazing --> [101, 2023,2607,2003, 6429, 999 ,102]    -->[-4.3630, 4.6859]    -->   POSITIVE:99.89%  NEGATIVE:0.,11%

# Preprocessing with a tokenizer

Like other neural networks, Transformer models can't process raw text directly, so the first step of our pipeline is to convert the text inputs into numbers that the model can make sense of. To do this we use a tokenizer, which will be responsible for:


*   splitting the input into words, subwords, or symbols(like punctuation) that are called tokens
*   Mapping each token to an integer
*   Adding additional inputs that may be useful to model



**Inputs IDs**        [101, 2023,2607,2003, 6429, 999 ,102]

**Special tokens**   [[CLS], this, course, is, amazing, !, [SEP]]

**Tokens**     [this, course, is, amazing, !]

**Raw text**  this course is amazing!

All this preprocessing needs to be done in exactly the same way as when the model was pretrained, so we first need to download that information from the Model Hub. To do this, we use the AutoTokenizer class and its from_pretrained( method.

In [ ]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english" # define the checkpoint that the model use
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

### The AutoTokenizer class can load the tokenizer for any checkpoint.

In [ ]:
raw_inputs = [
    "I've been waiting for a HuggingFace course my whole life.",
    "I hate this so much!",
]

inputs_ids = tokenizer(raw_inputs, padding= True, truncation=True,return_tensors="pt") # pt = pytorch is more suitable for transformers model , the padding fill the numbers by 0
print(inputs_ids)

{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]])}


In [ ]:
'''{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172, 2607,  2026,  2878,  2166,  1012,   102],
        [  101,  1045,  5223,  2023,  2061,  2172,   999,   102,     0,     0,  0,  0,     0,     0,     0,     0]]),
 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],   # allow of the model to understand that 1 is a meaningful word,
                           [1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]   #  0 is paddig
                           ]
                          )}'''


## Going through the model

We can download our pretrained model the same way we did with our tokenizer. Transformers library provides an AutoModel class which also has a from_pretrained() method:

In [ ]:
from transformers import AutoModel
checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModel.from_pretrained(checkpoint)

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

## AutoModel will download and cache the architecture, configuration and weights of the model.

In [ ]:
logits = model(**inputs_ids)
print(logits.last_hidden_state.shape)

torch.Size([2, 16, 768])


torch.Size([2,16,768]): raw_inputs has 2 sentences, the max length of a sentence is 16 (tokens), the last hidden layer has 768 neurons


*   **batch size** : 2
*  **sequence length** : 16
*   **hidden size** : 768



### There are many different model architectures available in Transformers, with each one designed around tackling a specific task. Here is a non-exhaustive list:


*   ForCausalLM
*   ForMaskedLM
*   ForMultipleChoice
*   ForQuestionAnswering
*   ForSequenceClassification
*   ForTokenClassification

For our example, we will need a model with a sequence classification head ( to be able to classify the sentences as positive or negative). So, we won't actually use the AutoModel class, but AutoModelForSequenceClassification:






In [ ]:
from transformers import AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
logits = model(**inputs_ids)
print(logits.logits.shape)

torch.Size([2, 2])


since we have just two sentences and two labels, the result we get from our model is of shape 2*2.

# Postprocessing the output

The values we get as output from our model don't necessarily make sense by themselves. Let's take a look:

In [ ]:
print(logits)

SequenceClassifierOutput(loss=None, logits=tensor([[-1.5607,  1.6123],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)


In [ ]:
print(logits.logits)

tensor([[-1.5607,  1.6123],
        [ 4.1692, -3.3464]], grad_fn=<AddmmBackward0>)


Our model predicted [-1.5607, 1.6123] for the first sentence and [4.1692, -3.3464] for the second one. Those are not probabilities but logits, the raw, unnormalized scores outputted by the last layer of the model. To be converted to probabilities, they need to go through a SoftMax layer.

In [ ]:
import torch

torch.set_printoptions(sci_mode = False)

predictions = torch.nn.functional.softmax(logits.logits, dim = -1)

print(predictions)

tensor([[0.0402, 0.9598],
        [0.9995, 0.0005]], grad_fn=<SoftmaxBackward0>)


Now we can see the model predicted [0.0402, 0.9598] for the first sentence and [0.9995, 0.005] for the second one. These are recognizable probability scores. To get the labels corresponding to each position, we can inspect the id2label attribute of the model config:

In [ ]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}

- first sentence: Negative 4.02% Positive 95.98%
- second sentence: Negative 99.46% Positive 0.54%

# Hugging Face Datasets

In [1]:
! pip install datasets

In [3]:
from datasets import load_dataset

ds = load_dataset("stanfordnlp/imdb")

README.md: 0.00B [00:00, ?B/s]

plain_text/train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

plain_text/test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

plain_text/unsupervised-00000-of-00001.p(…):   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

In [12]:
from datasets import load_dataset

# This command downloads and caches the dataset, by default in /.cache/huggingface/datasets.

raw_datasets = load_dataset("nyu-mll/glue", "mrpc")
raw_datasets

README.md: 0.00B [00:00, ?B/s]

mrpc/train-00000-of-00001.parquet:   0%|          | 0.00/649k [00:00<?, ?B/s]

mrpc/validation-00000-of-00001.parquet:   0%|          | 0.00/75.7k [00:00<?, ?B/s]

mrpc/test-00000-of-00001.parquet:   0%|          | 0.00/308k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/3668 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/408 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1725 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 3668
    })
    validation: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 408
    })
    test: Dataset({
        features: ['sentence1', 'sentence2', 'label', 'idx'],
        num_rows: 1725
    })
})

In [13]:
raw_datasets["train"]

Dataset({
    features: ['sentence1', 'sentence2', 'label', 'idx'],
    num_rows: 3668
})

In [14]:
raw_datasets["train"][6]

{'sentence1': 'The Nasdaq had a weekly gain of 17.27 , or 1.2 percent , closing at 1,520.15 on Friday .',
 'sentence2': 'The tech-laced Nasdaq Composite .IXIC rallied 30.46 points , or 2.04 percent , to 1,520.15 .',
 'label': 0,
 'idx': 6}

In [15]:
raw_datasets["train"][10:15]

{'sentence1': ['Legislation making it harder for consumers to erase their debts in bankruptcy court won overwhelming House approval in March .',
  'The Nasdaq composite index increased 10.73 , or 0.7 percent , to 1,514.77 .',
  'But he added group performance would improve in the second half of the year and beyond .',
  "He told The Sun newspaper that Mr. Hussein 's daughters had British schools and hospitals in mind when they decided to ask for asylum .",
  'Gyorgy Heizler , head of the local disaster unit , said the coach was carrying 38 passengers .'],
 'sentence2': ['Legislation making it harder for consumers to erase their debts in bankruptcy court won speedy , House approval in March and was endorsed by the White House .',
  'The Nasdaq Composite index , full of technology stocks , was lately up around 18 points .',
  'De Sole said in the results statement that group performance would improve in the second half of the year and beyond .',
  '" Saddam \'s daughters had British scho

The features attribute give us more information about a dataset splits columns

In [16]:
raw_datasets["train"].features

{'sentence1': Value('string'),
 'sentence2': Value('string'),
 'label': ClassLabel(names=['not_equivalent', 'equivalent']),
 'idx': Value('int32')}

Behind the scenes, label is of type ClassLabel, and the mapping of integers to label name is stored in the names folder. 0 corresponds to not_equivalent, and 1 corresponds to equivalent

# The mp Function

The map function allows you to apply a function over all the splits of a dataset.

In [20]:
from transformers import AutoTokenizer

checkpoint = "microsoft/Phi-3.5-mini-instruct"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

def tokenize_function(example):
  return tokenizer(example["sentence1"], example["sentence2"], padding="max_length", max_length = 128)

tokenized_datasets = raw_datasets.map(tokenize_function, batched = True)
print(tokenized_datasets.column_names)

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json:   0%|          | 0.00/306 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

Map:   0%|          | 0/3668 [00:00<?, ? examples/s]

Map:   0%|          | 0/408 [00:00<?, ? examples/s]

Map:   0%|          | 0/1725 [00:00<?, ? examples/s]

{'train': ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'attention_mask'], 'validation': ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'attention_mask'], 'test': ['sentence1', 'sentence2', 'label', 'idx', 'input_ids', 'attention_mask']}


In [22]:
tokenized_datasets = tokenized_datasets.remove_columns(["idx", "sentence1","sentence2"])
tokenized_datasets = tokenized_datasets.rename_column("label","labels")
tokenized_datasets = tokenized_datasets.with_format("torch")
tokenized_datasets["train"]

Dataset({
    features: ['labels', 'input_ids', 'attention_mask'],
    num_rows: 3668
})

# Pre-processing Pairs of Sentences

when passing two sentences, the tokenizer returns a new field called token_type_ids

it tells the model which tokens belong to the first sentence, and which ones belong to the second sentence.

In [24]:
from transformers import AutoTokenizer

checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

sequences = ["Hello! This is a sentence.", "This is another sentence."]

tokenizer(sequences, padding=True, return_tensors="pt")

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

{'input_ids': tensor([[ 101, 7592,  999, 2023, 2003, 1037, 6251, 1012,  102],
        [ 101, 2023, 2003, 2178, 6251, 1012,  102,    0,    0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 0, 0]])}

In [25]:
from transformers import AutoTokenizer

checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)


tokenizer("Hello! This is a sentence.", "This is another sentence.", padding=True, return_tensors="pt")

{'input_ids': tensor([[ 101, 7592,  999, 2023, 2003, 1037, 6251, 1012,  102, 2023, 2003, 2178,
         6251, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

# Dynamic Padding

If sentences in the same batch are of different lengths, we need to add padding tokens

In [26]:
sequences = ["This is the first sentence.",
             "This is the second one. It is longer.",
             "The third one is even longer! It has many words.",
             "This one is short."]

             This is   the first sentence .    pad pad pad pad pad  pad
             This is   the second  one    .    It  is longer . pad  pad
             The third one  is    even  longer  !  It  has many words .
             This one  is short    .     pad    pad pad pad pas pad pad

Does this strategy make sense when dealing with a huge dataset? Should we pad in accordance with the longest sentence in the entire dataset?

Pros:
- All rows in a datset will have same length.

Cons:
lots of batches will have useless columns with pad tokens only.

**Alternate:** Pad sentences at the time of batch creation only. This is called dynamic padding

Pros:
- All batches will have the smallest size possible.

Cons:
- Dynamic shapes don't work well on accelerators (GPUs, TPUs).

# Standard Padding

In [29]:
from datasets import load_dataset
from transformers import AutoTokenizer

raw_datasets = load_dataset("nyu-mll/glue","mrpc")

checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

def tokenize_function(example):
  return tokenizer(example["sentence1"], example["sentence2"], padding = "max_length",max_length=128)

tokenized_datasets = raw_datasets.map(tokenize_function,batched=True)
tokenized_datasets = tokenized_datasets.remove_columns(["idx", "sentence1","sentence2"])
tokenized_datasets = tokenized_datasets.rename_column("label","labels")
tokenized_datasets = tokenized_datasets.with_format("torch")

Map:   0%|          | 0/3668 [00:00<?, ? examples/s]

Map:   0%|          | 0/408 [00:00<?, ? examples/s]

Map:   0%|          | 0/1725 [00:00<?, ? examples/s]

In [30]:
from torch.utils.data import DataLoader

train_dataloader = DataLoader(tokenized_datasets["train"], shuffle=True, batch_size=16)

for step, batch in enumerate(train_dataloader):
  print(batch["input_ids"].shape) # fixed shape
  if step > 5:
    break

torch.Size([16, 128])
torch.Size([16, 128])
torch.Size([16, 128])
torch.Size([16, 128])
torch.Size([16, 128])
torch.Size([16, 128])
torch.Size([16, 128])


# Dynamic Padding

In [31]:
from datasets import load_dataset
from transformers import AutoTokenizer

raw_datasets = load_dataset("nyu-mll/glue","mrpc")

checkpoint = "bert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

def tokenize_function(example):
  return tokenizer(example["sentence1"], example["sentence2"],truncation = True)

tokenized_datasets = raw_datasets.map(tokenize_function,batched=True)
tokenized_datasets = tokenized_datasets.remove_columns(["idx", "sentence1","sentence2"])
tokenized_datasets = tokenized_datasets.rename_column("label","labels")
tokenized_datasets = tokenized_datasets.with_format("torch")

Map:   0%|          | 0/3668 [00:00<?, ? examples/s]

Map:   0%|          | 0/408 [00:00<?, ? examples/s]

Map:   0%|          | 0/1725 [00:00<?, ? examples/s]

In [34]:
from torch.utils.data import DataLoader
from transformers import DataCollatorWithPadding

data_collator = DataCollatorWithPadding(tokenizer)
train_dataloader = DataLoader(tokenized_datasets["train"], shuffle=True, batch_size=16, collate_fn=data_collator)

for step, batch in enumerate(train_dataloader):
  print(batch["input_ids"].shape) # fixed shape
  if step > 5:
    break

torch.Size([16, 72])
torch.Size([16, 74])
torch.Size([16, 74])
torch.Size([16, 100])
torch.Size([16, 96])
torch.Size([16, 77])
torch.Size([16, 75])


# The Trainer API

# Easily fine-tune(a pretrained model) or train (from scratch) transformer models